### ISSUE TITLE
Bug: OcropyClip: TypeError: function takes exactly 1 argument (2 given)

### ISSUE THREAD

**Issue by jbarth-ubhd** [cisocrgroup/ocrd_cis#72]
Workflow:

```sh
. /usr/local/ocrd_all/venv/bin/activate
export TMPDIR=/dwork/tmp
export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
ocrd-create-mets.xml
( /usr/bin/time ocrd process \
"olena-binarize -I OCR-D-IMG -O OCR-D-N1 -P impl wolf" \
"anybaseocr-crop -I OCR-D-N1 -O OCR-D-N2" \
"olena-binarize -I OCR-D-N2 -O OCR-D-N3 -P impl wolf" \
"cis-ocropy-denoise -I OCR-D-N3 -O OCR-D-N4 -P level-of-operation page" \
"cis-ocropy-deskew -I OCR-D-N4 -O OCR-D-N5 -P level-of-operation page" \
"tesserocr-segment-region -I OCR-D-N5 -O OCR-D-N6" \
"segment-repair -I OCR-D-N6 -O OCR-D-N7 -P plausibilize true" \
"cis-ocropy-deskew -I OCR-D-N7 -O OCR-D-N8 -P level-of-operation region" \
"cis-ocropy-clip -I OCR-D-N8 -O OCR-D-N9 -P level-of-operation region" \
"tesserocr-segment-line -I OCR-D-N9 -O OCR-D-N10" \
"cis-ocropy-clip -I OCR-D-N10 -O OCR-D-N11 -P level-of-operation line" \
"cis-ocropy-resegment -I OCR-D-N11 -O OCR-D-N12" \
"cis-ocropy-dewarp -I OCR-D-N12 -O OCR-D-N13" \
"calamari-recognize -I OCR-D-N13 -O OCR-D-OCR -P checkpoint /usr/local/ocrd_models/calamari/calamari_models-0.3/fraktur_historical/*.ckpt.json"

) >cmd.log 2>&1
```

Log:
```
02:57:04.632 INFO ocrd.task_sequence.run_tasks - Start processing task 'cis-ocropy-clip -I OCR-D-N8 -O OCR-D-N9 -p '{"level-of-op
eration": "region", "dpi": 0, "min_fraction": 0.7}''
Traceback (most recent call last):
  File "/usr/local/ocrd_all/venv/bin/ocrd", line 8, in <module>
    sys.exit(cli())
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/ocrd/cli/process.py", line 26, in pro
cess_cli
    run_tasks(mets, log_level, page_id, tasks, overwrite)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/ocrd/task_sequence.py", line 149, in 
run_tasks
    raise Exception("%s exited with non-zero return value %s. STDOUT:\n%s\nSTDERR:\n%s" % (task.executable, returncode, out, err))
Exception: ocrd-cis-ocropy-clip exited with non-zero return value 1. STDOUT:

STDERR:
02:57:06.605 INFO processor.OcropyClip - INPUT FILE 0 / P_00001
02:57:07.682 INFO processor.OcropyClip - Page "OCR-D-N8_00001" uses 300.000000 DPI
Traceback (most recent call last):
  File "/usr/local/ocrd_all/venv/bin/ocrd-cis-ocropy-clip", line 8, in <module>
    sys.exit(ocrd_cis_ocropy_clip())
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/ocrd_cis/ocropy/cli.py", line 33, in ocrd_cis_ocropy_clip
    return ocrd_cli_wrap_processor(OcropyClip, *args, **kwargs)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/ocrd/decorators/__init__.py", line 81, in ocrd_cli_wrap_processor
    run_processor(processorClass, ocrd_tool, mets, workspace=workspace, **kwargs)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/ocrd/processor/helpers.py", line 68, in run_processor
    processor.process()
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/ocrd_cis/ocropy/clip.py", line 131, in process
    background_image = Image.new('L', page_image.size, background)
  File "/usr/local/ocrd_all/venv/lib/python3.7/site-packages/PIL/Image.py", line 2613, in new
    return im._new(core.fill(mode, size, color))
TypeError: function takes exactly 1 argument (2 given)

Command exited with non-zero status 1
```

**Comment by bertsky** [comment]
Thanks for the report!

Looks like this is related to [this SO issue](https://stackoverflow.com/questions/60958127/torchvision-transforms-topilimage-typeerror-function-takes-exactly-1-argument) around [this GH issue](https://github.com/python-pillow/Pillow/issues/1808) of Pillow. Which version do you have? (`pip show pillow`)

**Comment by bertsky** [comment]
BTW your workflow has a flaw: You cannot use clipping _after_ any processor that already adds derived images (`AlternativeImage`) on the same hierarchy level. Since...

https://github.com/cisocrgroup/ocrd_cis/blob/a6f90abc43faabf86691d6d689b4ad4470a6c107/ocrd_cis/ocropy/clip.py#L77-L79

...and so the processor won't actually do anything but print warnings...

https://github.com/cisocrgroup/ocrd_cis/blob/a6f90abc43faabf86691d6d689b4ad4470a6c107/ocrd_cis/ocropy/clip.py#L152-L155

I have updated the [workflow guide wiki page](https://github.com/OCR-D/ocrd-website/wiki/Workflow-Guide-clipping) to better reflect this. (I distinctly remember documenting this elsewhere already, though...)

**Comment by jbarth-ubhd** [comment]
> Thanks for the report!
> 
> Looks like this is related to [this SO issue](https://stackoverflow.com/questions/60958127/torchvision-transforms-topilimage-typeerror-function-takes-exactly-1-argument) around [this GH issue](https://github.com/python-pillow/Pillow/issues/1808) of Pillow. Which version do you have? (`pip show pillow`)

```
(venv) jb@pers109:/usr/local/ocrd_all> pip show pillow
Name: Pillow
Version: 7.2.0
```

**Comment by jbarth-ubhd** [comment]
Workflow has been generated by script according to https://ocr-d.de/en/workflows a few months ago.

**Comment by jbarth-ubhd** [comment]
I'll have here a small script implementing roughly the steps 0..14 of https://ocr-d.de/en/workflows . Could you have a look at it, especially `after any processor that already adds derived images (AlternativeImage) on the same hierarchy level`

```perl
#!/usr/bin/perl
use strict;

sub setFlags {
	my $s=shift @_;

	my %store;
	for my $okv (grep { length $_ } split /\s+/, $s) {
		my($ok, $ov)=split /=/, $okv, 2;
		$store{$ok}=$ov;
	}
	for(my $i=0; $i<scalar(@_); $i+=2) {
		$store{$_[$i]}=$_[$i+1];
	}

	my $out="";
	while(my($x,$y)=each %store) {
		$out.=" $x=$y";
	}
	return substr($out, 1);
}

sub getFlag { # not in use
	my $s=shift @_;

	my %store;
	for my $okv (grep { length $_ } split /\s+/, $s) {
		my($ok, $ov)=split /=/, $okv, 2;
		$store{$ok}=$ov;
	}

	return $store{$_[0]};
}


sub step0 { my($w, $flags)=@_;
# Step 0: Image Enhancement (Page Level, optional)
	step1($w, $flags);
}

sub step1 { my($w, $flags)=@_;
# Step 1: Binarization (Page Level) 
# Many of the following processors require binarized images.
	step2($w."ocrd-olena-binarize -I OCR-D-aaa -O OCR-D-BIN\n", $flags); # TODO wolf
}

sub step2 { my($w, $flags)=@_;
# Step 2: Cropping (Page Level)
# We strongly recommend to execute this step if your images are not cropped already 
	for my $crop ("ocrd-anybaseocr-crop -I OCR-D-BIN -O OCR-D-CROP", "ocrd-tesserocr-crop -I OCR-D-BIN -O OCR-D-CROP") {
		step3($w."$crop\n", $flags);
	}
}

sub step3 { my($w, $flags)=@_;
# Step 3: Binarization (Page Level)
# For better results, the cropped images can be binarized again at this point or later on (on region level).
	my $I=getFlag($flags, "I");
	my $Ip1=$I+1;
	step4($w."ocrd-olena-binarize -I OCR-D-CROP -O OCR-D-BIN2\n", $flags); # TODO wolf
}

sub step4 { my($w, $flags)=@_;
# Step 4: Denoising (Page Level)
# This may not be necessary for all prints, and depends heavily on the selected binarization algorithm.
	for my $denoise ("ocrd-cis-ocropy-denoise -I OCR-D-BIN2 -O OCR-D-DENOISE",
			"ocrd-skimage-denoise -I OCR-D-BIN2 -O OCR-D-DENOISE") {
		step5($w."$denoise\n", $flags);
	}
}

sub step5 { my($w, $flags)=@_;
# Step 5: Deskewing (Page Level)
# The input images have to be binarized for this module to work.
	step6($w."ocrd-cis-ocropy-deskew -I OCR-D-DENOISE -O OCR-D-DESKEW-PAGE -P level-of-operation page\n", $flags);
}

sub step6 { my($w, $flags)=@_;
# Step 6: Dewarping (Page Level)
# The input image has to be binarized for the module to work.
# GPU REQUIRED!
	step7($w, $flags);
}

sub step7 { my($w, $flags)=@_;
# Step 7: Region segmentation
# Note: If you use ocrd-tesserocr-segment-region, which uses only bounding boxes instead of polygon coordinates, then you should post-process via ocrd-segment-repair with plausibilize=True to obtain better results without large overlaps.
# Note: The ocrd-sbb-textline-detector and ocrd-cis-ocropy-segment processors do not only segment the page, but also the text lines within the detected text regions in one step. Therefore with those (and only with those!) processors you don’t need to segment into lines in an extra step.
		for my $regSeg (
			"ocrd-tesserocr-segment-region -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-REG-TOBEREPAIRED && ocrd-segment-repair -I OCR-D-SEG-REG-TOBEREPAIRED -O OCR-D-SEG-REG -P plausibilize true",
			"ocrd-sbb-textline-detector -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-LINE -P model /path/to/model",
			"ocrd-cis-ocropy-segment -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-LINE -P level-of-operation page",
			"ocrd-anybaseocr-block-segmentation -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-REG -P block_segmentation_model /path/to/mrcnn -P block_segmentation_weights /path/to/model/block_segmentation_weights.h5",
			"ocrd-pc-segmentation -I OCR-D-DEWARP-PAGE -O OCR-D-SEG-REG") {
			step8($w."$regSeg\n", setFlags($flags, "lineSegDone", $regSeg=~/^(ocrd-sbb-textline-detector|ocrd-cis-ocropy-segment)/ ? 1 : 0));
		}
}

sub step8 { my($w, $flags)=@_;
# Step 8: Binarization (Region Level)
# If you already binarized your image twice on page level, and have no large images, you can probably SKIP this step.
	step9($w, $flags);
}

sub step9 { my($w, $flags)=@_;
# Step 9: Deskewing (Region Level)
	step10($w."ocrd-cis-ocropy-deskew -I OCR-D-BIN-REG -O OCR-D-DESKEW-REG -P level-of-operation region\n", $flags);
}

sub step10 { my($w, $flags)=@_;
# Step 10: Clipping (Region Level)
	step11($w."ocrd-cis-ocropy-clip -I OCR-D-DESKEW-REG -O OCR-D-CLIP-REG -P level-of-operation region\n", $flags);
}

sub step11 { my($w, $flags)=@_;
# Step 11: Line segmentation
# Note: If you use ocrd-tesserocr-segment-line, which uses only bounding boxes instead of polygon coordinates, then you should post-process with the processors described in Step 12. If you use ocrd-cis-ocropy-segment, you can directly go on with Step 13.
# Note: As described in Step 7, ocrd-sbb-textline-detector and ocrd-cis-ocropy-segment do not only segment the page, but also the text lines within the detected text regions in one step. Therefore with those (and only with those!) processors you don’t need to segment into lines in an extra step.

	if($flags=~/\blineSegDone=0\b/) {
		for my $lineSeg (
				"ocrd-cis-ocropy-segment -I OCR-D-CLIP-REG -O OCR-D-SEG-LINE -P level-of-operation region",
				"ocrd-tesserocr-segment-line -I OCR-D-CLIP-REG -O OCR-D-SEG-LINE") {
			step12($w."$lineSeg\n", setFlags($flags, "mustResegmentLine", $lineSeg=~/^ocrd-tesserocr-segment-line/ ? 1 : 0));
		}
	}
	else {
		step13($w, $flags);
	}
}

sub step12 { my($w, $flags)=@_;
	if($flags=~/\bmustResegmentLine=1\b/) {
		for my $resegLine (
				"ocrd-cis-ocropy-clip -I OCR-D-SEG-LINE -O OCR-D-CLIP-LINE -P level-of-operation line",
				"ocrd-cis-ocropy-resegment -I OCR-D-SEG-LINE -O OCR-D-RESEG") {
			step13($w."$resegLine\n", $flags);
		}
	}
	else {
		step13($w, $flags);
	}
}

sub step13 { my($w, $flags)=@_;
	step14($w."ocrd-cis-ocropy-dewarp -I OCR-D-CLIP-LINE -O OCR-D-DEWARP-LINE\n", $flags);
}

sub step14 { my($w, $flags)=@_;
	for my $ocr (
			'TESSDATA_PREFIX="/test/data/tesseractmodels/" ocrd-tesserocr-recognize -I OCR-D-DEWARP-LINE -O OCR-D-zzz -P model Fraktur',
			"ocrd-calamari-recognize -I OCR-D-DEWARP-LINE -O OCR-D-zzz -P checkpoint /path/to/models/\*.ckpt.json") {
		printStep($w."$ocr\n", $flags);
	}
}

sub printStep { my ($w, $flags)=@_;

	my $n=3;
	$w=~s!(-[IO] OCR-D-)([A-Z0-9-]+)\b!sprintf("%s%03d", $1, int($n++/2))!ge;
	$w=~s!OCR-D-aaa!OCR-D-IMG!g;
	$w=~s!OCR-D-zzz!OCR-D-OCR!g;

	print "$w # $flags\n---\n";
}

step0("", "");
```

**Comment by jbarth-ubhd** [comment]
let's try Pillow 7.1.2...
```
ocrd-utils 2.17.2 requires Pillow>=7.2.0, but you'll have pillow 7.1.2 which is incompatible.
Successfully installed pillow-7.1.2
```

**Comment by jbarth-ubhd** [comment]
> You cannot use clipping after any processor that already adds derived images

this two lines? 
```
cis-ocropy-deskew -I OCR-D-N7 -O OCR-D-N8 -P level-of-operation region" \
cis-ocropy-clip ...
```

But it's still here: https://ocr-d.de/en/workflows Step 9 + 10

**Comment by bertsky** [comment]
> > You cannot use clipping _after_ any processor that already adds derived images
> 
> this two lines?
> 
> ```
> cis-ocropy-deskew -I OCR-D-N7 -O OCR-D-N8 -P level-of-operation region" \
> cis-ocropy-clip ...
> ```
> 
> But it's still here: https://ocr-d.de/en/workflows Step 9 + 10

Yes, sorry, did not spot it there the first time. I have updated [my issue](https://github.com/OCR-D/ocrd-website/issues/172) on the recommendations.

**Comment by bertsky** [comment]
> > Thanks for the report!
> > Looks like this is related to [this SO issue](https://stackoverflow.com/questions/60958127/torchvision-transforms-topilimage-typeerror-function-takes-exactly-1-argument) around [this GH issue](https://github.com/python-pillow/Pillow/issues/1808) of Pillow. Which version do you have? (`pip show pillow`)
> 
> ```
> (venv) jb@pers109:/usr/local/ocrd_all> pip show pillow
> Name: Pillow
> Version: 7.2.0
> ```

7.2.0 should have the fix IIUC. I am still trying to reproduce. It definitely depends on the image's color mode (16/32 bit integer or float are probably the culprit). Getting back to you when I tried with such an image.

**Comment by bertsky** [comment]
> I am still trying to reproduce. It definitely depends on the image's color mode (16/32 bit integer or float are probably the culprit). Getting back to you when I tried with such an image.

No luck here. Tried with 8+1, with 16 bit and 32 bit images. Could you please share the image it failed on?

**Comment by jbarth-ubhd** [comment]
Sent an email with link to images.

**Comment by bertsky** [comment]
Thanks – I could reproduce. You helped find a bug again! Fix is in #75
